### ISSUE TITLE
Worse then linear scaling with time

### ISSUE THREAD

**Issue by JamiePringle** [OceanParcels/parcels#1064]
I have been using ocean parcels to track particles released from small locations within the Mercator 1/12 flow fields, and I am finding that when the duration of the run increases, the run time increases much more then I would expect.  For example, when I increase the integration time from 12 days to 60 days, a factor of 5, the run time increases by a factor of 10. This issue becomes much worse when I try to, for example, model larval release over a season.  This effect is making it dramatically harder to do my analyses, and is making me start to think about stoping and restarting each run after about 10 days to improve the performance; this would be cumbersome and inconvenient.

Do you understand why the scaling is worse than the expected linear, and how I might fix it? Is this expected behavior?

I am running parcels 2.2.2 on python 3.6.13

The top panel of the attached figure shows the run-time as a function of the duration of the simulation of a single release of a large number of particles in a small area on a single day. The dashed line shows the expected linear scaling.  The bottom panel shows the time it takes to integrate a day as a function of the model-day. You can see quite dramatically as the run goes on the time to calculate each day increases.  All disk caches have been cleared before each run with “sync; echo 2 > /proc/sys/vm/drop_caches”
![scalingFigure](https://user-images.githubusercontent.com/12818667/126354788-3ad288fa-6716-434c-a9f4-5aeb7784bc25.png)

I have spent some time trying to eliminate some possible issues.  The input fields are global 1/12 degree Mercator/NEMO runs. The data is in compressed netCDF fields chunked by 1 in the vertical and 256 by 256 in the horizontal. There is one file/day. I am using the standard netCDF4 engine in the field set, and chunking by 1,512,512, which I have found to be optimal.  Most runs are made with 144 particles at each depth released at 8 depths in the top few meters in a horizontally dense patch. To make sure the issue is not related to the initial particle patch expanding so that more chunks of data are read in,  the particles are released so closely in depth and horizontally that they do not disperse by more than 1 degree in the longest integration (well within 1 netcdf chunk which is about 20 degrees of latitude squared). If the initial particle distribution is larger, the issue remains. 

The effect is seen when running in both MPI and in scaler mode, and is not sensitive to the number of MPI processes. The output frequency is daily, so the output load will scale linearly and steadily with the number of days of integration. 

I would be happy to take any suggestions, or to provide model code or details to illustrate the problem. The code that launches the particles is included. 

Do you have any ideas on why this is occurring?

Thanks,
Jamie Pringle

[trackPointsScript.zip](https://github.com/OceanParcels/parcels/files/6849718/trackPointsScript.zip)
Jamie Pringle
Ocean Process Analysis Laboratory in the Institute for Earth, Ocean and Space
Professor of Earth Sciences
University of New Hampshire
http://oxbow.sr.unh.edu

**Comment by JamiePringle** [comment]
Ok, the attached code in the message above is too complex to be really useful for y'all. Here is a simpler version, made with the example code in https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_periodic_boundaries.ipynb.

The code is in the attached zip file, and can be run with the data from the ocean parcels example for the periodic boundary. 

This figure, which is produced by the code, shows the time it takes to run the particle tracking devided by the length of the model integration in days. A straight line is drawn to show a linear scaling for comparison. Roughly, by eye, it looks quadratic, which suggest that an O(N^2) algorithm snuck into the code somewhere. 

I hope this simple example helps. 

Jamie Pringle
![ScalingResultSimple](https://user-images.githubusercontent.com/12818667/126397950-4d5a64fa-68b3-4086-bedf-6a04b4638b96.png)
[parcelsTest_analytic.zip](https://github.com/OceanParcels/parcels/files/6851313/parcelsTest_analytic.zip)

**Comment by erikvansebille** [comment]
Thanks @JamiePringle, for providing this second script with a minimal example. This also shows the issue is likely not related to input or output. I just ran it on my own local system and got very similar worse-than-linear results... 

Your script is a very good starting point for us to dig into this scaling behaviour. Note that @CKehl has written a paper a few months ago (still under review, but in preprint at https://arxiv.org/abs/2105.00057) that also includes scaling analysis of some of the new ParticleSet-datastructures he's implemented in Parcels. I'm sure he'll have something to say about your analysis and graph too

We'll stay in touch as we investigate what's going on!

**Comment by erikvansebille** [comment]
OK, I already made a first important observation. If you remove the `output_file=output_file` from the `pset.execute()`, then the scaling is _much_ better! (note that the y-axis range is tiny)
<img width="631" alt="Screen Shot 2021-07-21 at 08 44 55" src="https://user-images.githubusercontent.com/14315062/126443688-ced6737f-ab80-4afc-ad53-a7123bfd7c34.png">


Note that, in this case, the entire simulation is done in JIT (there is no need to exit JIT to call `ParticleSet.write()`)

This suggests that the reason for the worse-than-linear performance is either
1. In the ParticleSet.write() method
2. In the entering/exiting of JIT

I'll dig a bit deeper!

**Comment by erikvansebille** [comment]
And another update: I have now done a simulation where I replaced the particle writing with a `callbackdt=delta(hours=1)` to trigger the entering-and-exiting of JIT. See image below. 
![ScalingResultSimple](https://user-images.githubusercontent.com/14315062/126458576-264a5866-f8ec-4c5e-bbf3-53668a930626.png)

The runtime _per day_ is constant, so that is good! This suggests that indeed the problem is in the writing of the ParticleFile; I'll dig deeper!

**Comment by CKehl** [comment]
Hi.
So, general comments first:
a) good that you measure it in a Python file, not a notebook - jupyter makes runtime measurements unreliable due to its internal caching.
b) using the `time` module to make runtime measurements is not reliable. I gave an extensive talk about runtime measurements in our group (-> Python for Lunch), which in particular shows how unreliable MPI runtimes are with the `time` / `deltatime` module. Good approach to get some first numbers, but bad idea to derive performance conclusions from. Best suggestion: use the `perf_counter` (https://docs.python.org/3/library/time.html#time.perf_counter)

Now on the actual observation (even though, as said, the measurement needs to be confirmed by proper timers): the bottleneck here is the write procedure and we are aware of that for already ~8 months. I can remember a similar issues at the end of 2020 here on github by a parcels user who ran out-of-memory during the write-procedure of the particle set. The way writes are done is an exponential runtime behaviour, because all time step files need to be parsed at each write occurrence to determine the lasted timestep written. Obviously, the more timesteps have been written, the higher the listing, parsing and evaluation effort for `psets`. Sadly, the way writes are done is quite fixed with how the kernel main loop works. Iterative writes of particles into a files would be great, but doesn't work with NetCDF and the CF Standard for the oceanic data that we need to comply with.

**TL;DR**: if it was as easy as just writing all particles to file at each `outdt` and we could rewrite the whole file writing to use HDF5 and appending dimensions, then that behaviour lessened / minimized. But: because we also need to write particles on kernel errors and because we need to parse the last written timestep from file, the writing procedure is more complex than one may think. There are no plans to redesign or rewrite the particle file interface in the near-term future, to the best of my knowledge. And in the end, the writing procedure will still - in the very best of circumstances - not be entirely constant. And its going to slow-down the simulation anyway considerably.

@erikvansebille see `parcels/particlefile/particlefilesoa.py`, function `read_from_npy`, l. 63-96 (in my branch checkout - line numbers may vary for you); `for npyfile in file_list:` loops over all written files - the longer the simulation, the more files need to be parsed.

**Comment by erikvansebille** [comment]
OK, I think I've found the culprit. It's the lines https://github.com/OceanParcels/parcels/blob/5d5a25957fab66d99d6b28f9aa794061a42e2fe0/parcels/particlefile/baseparticlefile.py#L299-L300

Note that this is only the _writing_ of the `pset_info.npy` file. If I replace line 300 it with a `pass`, I get linear behaviour in the test script developed by @JamiePringle. 

It does make sense that the writing of this `pset_info.npy` file scales poorly, as it (among other variables) contains the times written in all temporary `npy` dumps of the particleset. This list grows on every call to `pset.write()` and is needed to later combine all the temporary `npy` dumps into one netcdf file. Indeed, I also get linear (good) scaling behaviour if I remove `time_written` from the list of variables to be written in 
https://github.com/OceanParcels/parcels/blob/5d5a25957fab66d99d6b28f9aa794061a42e2fe0/parcels/particlefile/particlefilesoa.py#L58-L60

I had not expected the repeated writing of this slowly growing file to be so detrimental to performance, though. Although I now realise that, for the 400 days integration the size of this file is 533KB for 8000 snapshots, the total amount of data written just for this file is approximately 800*0.5MB/2=200MB, which is a significant amount of writing yes!

The question then is how to fix this. One option would be to write the `pset_info.npy file` only at the end of the `pset.execute` (instead of on every call to `pset.write()`), but the downside is that it would be much more difficult to create a netcdf file from a not-completed simulation. 
Alternative is to compute the `time_written` from the local npy` dumps of the particleset; this should be possible. I'll try to make this work in a PR, and hope that this improves the poor scaling!

**Comment by erikvansebille** [comment]
OK, I think I fixed this Issue in #1066. See below the new scaling graph with the original script that can be compared to the one in https://github.com/OceanParcels/parcels/issues/1064#issuecomment-883928432. So I think it works now!

![ScalingResultSimple](https://user-images.githubusercontent.com/14315062/126488193-e223338b-a60f-44fe-afc4-6ac90eb1c277.png).

**Comment by JamiePringle** [comment]
Thanks! @erikvansebille Can I install and run this change by following the directions on the ocean parcels web page starting "installing a non-released version of parcels"?

@CKehl Thanks -- all of my timings were made by running the python code from the bash shell and seeing how long it took the command to return. I have played many benchmarking games :-)  It is only the simple example code that used time.time()

For both of you: I have just been at two seminars/meetings in which the main issue with using oceanParcels is its slow speed at big problems. For problems without behavior, people are putting up with the pain of Ariane because the speed of oceanParcels was much too slow.  We were puzzled because IO speed was not limiting for longer integrations, which ran against our experience with other packages.

To give you an example from my work: the reason that the first code I sent you is designed to be called as a script which reads a parameter file is that I had to refactor my code to run each larval release date as a separate job in parallel then to run them in one job. This caused a factor of 8 reduction in run time despite reading all the model data 5 times more.... Several colleagues tried parcels on small tests, were very happy with it, then were dismayed when they tried to do longer runs or runs with more release times. 

So this is a big deal, and thanks!

Jamie

**Comment by erikvansebille** [comment]
Thanks for this honest feedback, @JamiePringle! This is extremely useful to hear, and luckily @CKehl is currently working on an extensive study of the performance of Parcels in different 'real-world' experiments, including the simulations done for our work on [sinking foraminifera](https://doi.org/10.1029/2019PA003606), on [sinking plastics](https://doi.org/10.1029/2020JC017098) and a few others. Hopefully, this will also help us to specifically target some of these performance issues for larger runs...

As for how to install: yes follow the developer instructions and then choose the `fixing_psetinfo_performance_problem` branch. Would be keen to hear if it also works for your real-world simulations

**Comment by erikvansebille** [comment]
For completeness, here is the comparison of the total runtime for the original `master` on the top, and the new PR #1066 on the bottom. The total runtime for the 400-day simulation reduces from ~100 seconds to ~20 seconds, and the latter is also much more linear, as hoped.
![ScalingResultSimple_orig](https://user-images.githubusercontent.com/14315062/126497732-1070c7e5-2a88-472a-867c-2bcadc039a89.png)
![ScalingResultSimple_1066](https://user-images.githubusercontent.com/14315062/126497749-ff132eb4-0661-49a4-8fb4-8970f559358f.png)

**Comment by JamiePringle** [comment]
Erik and Christian -- I have done my tests of the code, and the scaling is indeed linear in my Mercator 1/12 case, and the results are consistent with my prior results (but are returned much, much faster).

Thanks!

I will leave it to you to close this issue if you think you are done with it. I think this change will be very useful to many people. 

Jamie

**Comment by erikvansebille** [comment]
Ah, that is so good to here @JamiePringle! Really glad that you brought up this issue and that we were able to fix it with #1066. 

Before I close the issue, could you provide (rough) timings of the speed improvement you experienced in the Mercator 1/12 case? Would be good to know

**Comment by JamiePringle** [comment]
Yes, will post after my lunch time (meeting with students now). Should be
up when you see it tomorrow your time.

**Comment by JamiePringle** [comment]
Just to close the loop, here is a comparison of the old code and the new code in the same format as the my plot from start of this thread. The grey dashed line shows what would be expected for linear scaling of run time to integration duration.

You can see the new code does very well. It is both faster and has linear scaling with increasing run duration.

The top panel of the attached figure shows the run-time as a function of the duration of the simulation of a single release of a large number of particles in a small area on a single day. The dashed line shows the expected linear scaling. The bottom panel shows the time it takes to integrate a day as a function of the model-day.

![scalingFigure](https://user-images.githubusercontent.com/12818667/126521772-5720d577-edde-4b79-801d-ee0e279181d3.png)

Cheers,
Jamie

**Comment by erikvansebille** [comment]
Thanks Jamie, this is extremely useful info; good to see that this fix #1066 works! I will now merge that PR and close this Issue
